aptos-move/framework/src/module_metadata.rs

```diff
@@ -91,6 +95,66 @@ pub fn get_module_metadata(module: &CompiledModule) -> Option<RuntimeModuleMetad
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Error)]
+#[error("Unknown attribute ({}) for key: {}", self.attribute, self.key)]
+pub struct MetadataValidationError {
+    pub key: String,
+    pub attribute: u16,
+}
+
+pub fn is_valid_view_function(
+    functions: &BTreeMap<Identifier, Function>,
+    fun: &str,
+) -> Result<(), MetadataValidationError> {
+    if let Ok(ident_fun) = Identifier::new(fun) {
+        if let Some(mod_fun) = functions.get(&ident_fun) {
+            if !mod_fun.return_.is_empty() {
+                return Ok(());
+            }
+        }
+    }
+
+    Err(MetadataValidationError {
+        key: fun.to_string(),
+        attribute: KnownAttributeKind::ViewFunction as u16,
+    })
+}
+
+pub fn verify_module_metadata(module: &CompiledModule) -> Result<(), MetadataValidationError> {
+    let metadata = if let Some(metadata) = get_module_metadata(module) {
+        metadata
+    } else {
+        return Ok(());
+    };
+
+    let functions = module
+        .function_defs
+        .iter()
+        .map(|func_def| Function::new(module, func_def))
+        .collect::<BTreeMap<_, _>>();
+    for (fun, attrs) in &metadata.fun_attributes {
+        for attr in attrs {
+            if attr.is_view_function() {
+                is_valid_view_function(&functions, fun)?
```

Comment:
You should add a comment here or at this function that you only validate partial correctness of view functions here. The more heavy weight part is happening at execution time, including validation of actual arguments. With the design you are suggesting here, it would probably be cleaner to pull all static verification out of `aptos-vm/src/verifier` and into here, also for entry functions, but not in this PR...